orttraining/orttraining/core/graph/gradient_builder_base.h

```diff
@@ -85,33 +87,57 @@ class GradientBuilderBase {
     return gradient_graph_config_;
   }
 
+  void TryRecordStashedForwardGraphTensor(const std::string& name) const {
+    if (stashed_tensors_.find(name) == stashed_tensors_.end()) {
+      stashed_tensors_.insert(name);
+    }
+  }
+
+  bool IsTensorStashed(const std::string& name) const {
+    return stashed_tensors_.find(name) != stashed_tensors_.end();
+  }
+
   // i-th input of forward op
-  ArgDef I(const size_t i) const {
+  ArgDef I(const size_t i, bool record_stashing = true) const {
     ORT_ENFORCE(i < node_->InputDefs().size());
 
     const std::string& name = node_->InputDefs()[i]->Name();
     const NodeArg* recomputed_nodearg = graph_->GetNodeArg(graph_utils::RecomputeName(name));
     if (recomputed_nodearg) {
       const Node* producer_node = graph_->GetProducerNode(name);
       LOGS(logger_, INFO) << "Recomputed node arg found for " << producer_node->Name();
+      if (record_stashing) {
+        TryRecordStashedForwardGraphTensor(recomputed_nodearg->Name());
+      }
```

Comment:
Do you see problem if we share this two checks at line 109 and 116 by moving the checks to line 105. ?

Reply:
Same for O()